ens-improvement-proposals/ens-l2-offchain.md

```diff
@@ -0,0 +1,117 @@
+
+**EIP 3668** allows for offchain (including Layer 2/L2) lookups of data in a way that is transparent to clients and provides contract authors to implement whatever validation is necessary; in many cases, this can be provided without any additional trust assumptions over and above those required if data is stored onchain.
+
+**ENSIP 10** is a general way to resolve wildcard (eg: *.foo.eth) on L1. Issuing subdomains and moving the resolution of the parent name offchain allows dapps to create subdomains offchain yet make it accessible through L1.
+
+**ENSIP 11** allows a single name to hold different addresses across multiple EVM compatible chains. Contract addresses can use these EVM chain addresses so that clients can choose appropriate addresses on different chains.
+
+## Steps required for Dapps and wallets to support offchain data lookup.
+
+If your dapps or wallets use one of those libraries, the EIP 3668 and ENSIP 10 support will be built in, so simply update the library when ready.
+
+- [ethersjs](https://github.com/ethers-io/ethers.js)
+- [web3js](https://github.com/ethereum/web3.js)
+
+### ethersjs
+
+Currently EIP 3668 is implemented as npm module as [@chainlink/ethers-ccip-read-provider
+](@chainlink/ethers-ccip-read-provider
+)([source](https://github.com/smartcontractkit/ccip-read/tree/rewrite/packages/ethers-ccip-read-provider)).
+
+The basic usage example is as follows.
+
+```js
+import { CCIPReadProvider } from '@chainlink/ethers-ccip-read-provider';
+import { abi as IExtendedResolver_abi } from '@ensdomains/offchain-resolver-contracts/artifacts/contracts/IExtendedResolver.sol/IExtendedResolver.json';
+const IExtendedResolver = new ethers.utils.Interface(IExtendedResolver_abi);
+const baseProvider = ethers.getDefaultProvider(options.provider);
+const provider = new CCIPReadProvider(baseProvider);
+```
+
+The wildcard is still not supported yet, but one can easily be implemented as follows.
+
+```js
+const labels = name.split('.');
+let resolverAddress = undefined;
+for(let i = 0; i < labels.length; i++) {
+    resolverAddress = await registry.resolver(ethers.utils.namehash(labels.slice(i).join('.')));
+    if(resolverAddress !== "0x0000000000000000000000000000000000000000") {
+    break;
+    }
+}
+if(resolverAddress === undefined) {
+    console.log(`${name} could not be resolved`);
+}
+
+const data = Resolver.encodeFunctionData('addr(bytes32)', [node]);
+const responseData = await resolver.resolve(dnsName(name), data);
+const addr = Resolver.decodeFunctionResult('addr(bytes32)', responseData);
+```
+
+Please refer to [offchain resolver client example code](https://github.com/ensdomains/offchain-resolver/blob/main/packages/client/src/index.ts#L46) for more detail.
+
+### web3js
+
+Work in progress
+
+### Any other libraries
+
+If you use other libraries or custom integration, you have to add the support on your own. So please read each specification and support accordingly or contact the ENS team for help.
```

Comment:
I would lead by saying they should open an issue for support so we can track efforts to add 3668/ENSIP10 support to other libraries.